File: Tools/flashtool/DockWindows/ConsoleWindow.py
from ConsoleWindow_ui import *
from msg_structs import *
from Dispatcher import dispatcher
from AnsiParser import AnsiParser

from PyQt4.QtCore import *
from PyQt4.QtGui import *

import sys
from datetime import datetime
from Queue import Queue

ansi_palette = {
    0:  QBrush(QColor("#000")),
    1:  QBrush(QColor("#C00")),
    2:  QBrush(QColor("#0C0")),
    3:  QBrush(QColor("#CC0")),
    4:  QBrush(QColor("#00C")),
    5:  QBrush(QColor("#C0C")),
    6:  QBrush(QColor("#0CC")),
    7:  QBrush(QColor("#CCC")),
    8:  QBrush(QColor("#444")),
    9:  QBrush(QColor("#F44")),
    10: QBrush(QColor("#4F4")),
    11: QBrush(QColor("#FF4")),
    12: QBrush(QColor("#44F")),
    13: QBrush(QColor("#F4F")),
    14: QBrush(QColor("#4FF")),
    15: QBrush(QColor("#FFF"))
}


test = """
\x1B[0m   _  \x1B[1;31m_\x1B[0m       \x1B[1;31m______\x1B[0m                             \x1B[1;31m______\x1B[0m
  / \x1B[1;30m\\\x1B[31m/\x1B[0m \x1B[31m\\\x1B[1;33m_\x1B[0m    \x1B[1;31m/\x1B[0m \x1B[31m____/\x1B[37m_______  __  ______   _  __ \x1B[1;31m/\x1B[0;31m_\x1B[37m  \x1B[31m__/\x1B[37m______   __  ___
 /  \x1B[1;31m/ \x1B[0m \x1B[1;33m/\x1B[0m \x1B[33m\\\x1B[37m  \x1B[1;31m/\x1B[0m \x1B[31m/\x1B[37m \x1B[1;31m__ \x1B[0m/ \x1B[1;30m__\x1B[0m/ _ \x1B[1;30m\\\x1B[0m/  \x1B[1;30m|\x1B[0m/  / _ \x1B[1;30m|\x1B[0m / \x1B[1;30m|\x1B[0m/ \x1B[1;30m/\x1B[0m  \x1B[1;31m/\x1B[0m \x1B[31m/\x1B[37m / \x1B[1;30m__\x1B[0m/ _\x1B[1;30m |\x1B[0m /\x1B[1;30m  |\x1B[0m/\x1B[1;30m  /\x1B[0m
 \\\x1B[1;30m_/\x1B[31m\\\x1B[0;31m_\x1B[1;33m/\x1B[0m  \x1B[33m/\x1B[37m \x1B[1;31m/\x1B[0m \x1B[31m/\x1B[1m_/\x1B[0m \x1B[31m/\x1B[37m/ \x1B[1;30m_/\x1B[0m/ \x1B[1;30m, _\x1B[0m/\x1B[1;30m /|_/ \x1B[0m/\x1B[1;30m __ |\x1B[0m/\x1B[1;30m    /\x1B[0m \x1B[1;31m /\x1B[0m \x1B[31m/\x1B[37m / \x1B[1;30m_/\x1B[0m/ \x1B[1;30m__\x1B[0m \x1B[1;30m|\x1B[0m/ \x1B[1;30m/|_/\x1B[0m \x1B[1;30m/\x1B[0m
      \x1B[1;33m\\\x1B[0;33m_/\x1B[37m  \x1B[1;31m\\\x1B[0;31m____/\x1B[37m/\x1B[1;30m___\x1B[0m/\x1B[1;30m_/\x1B[0m|\x1B[1;30m_\x1B[0m/\x1B[1;30m_/\x1B[0m  /\x1B[1;30m_\x1B[0m/\x1B[1;30m_/\x1B[0m |\x1B[1;30m_\x1B[0m/\x1B[1;30m_/|_/\x1B[0m  \x1B[1;31m/\x1B[0;31m_/\x1B[37m /\x1B[1;30m___\x1B[0m/\x1B[1;30m_/ \x1B[0m|\x1B[1;30m_\x1B[0m/\x1B[1;30m_/  \x1B[0m/\x1B[1;30m_/\x1B[0m
\x1B[0m

\x1B[0m\x1B[1m    \x1B[0m   \x1B[1m_\x1B[32m_\x1B[0;37m           \x1B[1m_\x1B[32m_\x1B[0;37m       \x1B[1m__\x1B[32m__\x1B[0;37m      \x1B[1m_\x1B[32m_\x1B[0;37m  \x1B[1m_\x1B[32m_\x1B[0;37m  \x1B[1;30m     \x1B[0;32m \x1B[1;30mf00^06\x1B[0;37m
\x1B[1;30m  \x1B[37m_\x1B[32m__\x1B[0;37m \x1B[1m/\x1B[0m \x1B[1;30m/\x1B[37m_\x1B[32m_\x1B[37m_\x1B[32m__ \x1B[37m_\x1B[32m__ \x1B[37m/\x1B[0m \x1B[1;30m/\x1B[32m \x1B[37m_\x1B[32m__ \x1B[0;37m \x1B[1m/\x1B[32m \x1B[37m/\x1B[0m \x1B[1;30m/\x1B[37m_\x1B[32m__ \x1B[37m_/\x1B[0m \x1B[1;30m/\x1B[32m_\x1B[37m/\x1B[0m \x1B[1;30m/\x1B[0;37m   \x1B[1m__\x1B[32m__\x1B[37m_\x1B[32m_ __\x1B[0;37m
 \x1B[1m(\x1B[30m_\x1B[32m-\x1B[30m<\x1B[32m/\x1B[0;37m  \x1B[1;32m'_\x1B[37m/\x1B[0m \x1B[1;32m-\x1B[30m_\x1B[32m)\x1B[0;37m \x1B[1;32m-\x1B[30m_)\x1B[0;37m \x1B[1;32m_\x1B[0;37m \x1B[1;32m\\\x1B[0;37m \x1B[1;32m_\x1B[0;37m \x1B[1;30m`\x1B[32m/ /\x1B[0;37m \x1B[1;30m/\x1B[37m/\x1B[32m _\x1B[0;37m \x1B[1;30m`\x1B[32m/ __/\x1B[0;37m \x1B[1;30m_\x1B[0;37m \x1B[1;32m\\\x1B[37m_/\x1B[0m \x1B[1;32m__\x1B[30m/\x1B[32m\\ \\\x1B[0;37m \x1B[1;30m/\x1B[0;37m
\x1B[1;32m/\x1B[30m___\x1B[32m/\x1B[30m_/\\_\\\\__/\\__\x1B[32m/\x1B[30m_.__/_,_\x1B[32m/\x1B[30m_\x1B[32m/\x1B[30m_\x1B[32m(\x1B[30m_)_,_/\\__\x1B[32m/\x1B[30m_/\x1B[32m/\x1B[30m_\x1B[32m(\x1B[30m_)__/\x1B[32m/\x1B[30m_\\_\\\x1B[0;37m
\x1B[0m
"""


# TODO:
# * http://doc.qt.digia.com/4.6/debug.html
#   qInstallMsgHandler() for qWarning(), qDebug()
# * Text Input
#
class ConsoleWindow(QMainWindow, Ui_ConsoleWindow):
    def __init__(self):
        QMainWindow.__init__(self)
        self.setupUi(self)

        self.setContextMenuPolicy(Qt.NoContextMenu)

        self.action_save.triggered.connect(self.action_save_triggered)
        self.action_clear.triggered.connect(self.action_clear_triggered)
        self.action_wrap.triggered.connect(self.action_wrap_triggered)

        self.cursor = QTextCursor(self.plainTextEdit.document())

        self.ansi_format = QTextCharFormat()

        self.ansi_parser = AnsiParser()
        self.ansi_parser.attrib_changed = self.ansi_attrib_changed
        self.ansi_parser.print_text = self.ansi_print_text

        self.queue = Queue()
        dispatcher.subscribe(self.queue, msg_shell_to_pc.MSG_ID)

        self.timer = QTimer()
        self.timer.timeout.connect(self.timer_timeout)
        self.timer.start(100)

    def action_save_triggered(self):
        t = datetime.now()

        fn = "log_%02d%02d%02d_%02d%02d%02d.txt" % (
            t.year % 100, t.month, t.day,
            t.hour, t.minute, t.second
        )

        modifiers = QApplication.keyboardModifiers()

        if modifiers != Qt.ShiftModifier:
            fn = str(QFileDialog.getSaveFileName(
                self, "Save console output", fn,
                "Text files (*.txt);;All Files (*.*)"
            ))

        if fn == "":
            return

        with open(fn, "w") as f:
            f.write(self.plainTextEdit.toPlainText())

    def action_clear_triggered(self):
        self.plainTextEdit.clear()

    def action_wrap_triggered(self):
        if self.action_wrap.isChecked():
            self.plainTextEdit.setLineWrapMode(QPlainTextEdit.WidgetWidth)
        else:
            self.plainTextEdit.setLineWrapMode(QPlainTextEdit.NoWrap)

    def ansi_print_text(self, text):
        self.cursor.insertText(text)

    def ansi_attrib_changed(self):
        fg = self.ansi_parser.foreground + (8 if self.ansi_parser.bold else 0)
        bg = self.ansi_parser.background
        
        self.ansi_format.setForeground(ansi_palette[fg])
        self.ansi_format.setBackground(ansi_palette[bg])

        self.cursor.setCharFormat(self.ansi_format)

    def timer_timeout(self):
        if self.action_pause.isChecked():
            return

        # combine messages
        #
        text = ""
        # text = "\x1b[31mRed\x1b[32mGreen\x1b[34mBlue"
        # text = test
        while not self.queue.empty():
            msg = self.queue.get()
            text += msg[2:]

        if text == "":
            return

        pte = self.plainTextEdit
        vsb = pte.verticalScrollBar()
        at_bottom = vsb.value() == vsb.maximum()

        pte.setUpdatesEnabled(False)

        self.ansi_parser.parse(text)
        sys.stdout.write(text)

        if at_bottom:
            vsb.setValue(vsb.maximum())

        pte.setUpdatesEnabled(True)

File: Tools/quadcontrol/DockWindows/ConsoleWindow.cpp
/**
 * Copyright (C)2015 Thomas Kindler <mail_drquad@t-kindler.de>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "ConsoleWindow.h"
#include "ui_ConsoleWindow.h"

#include <QSerialPort>
#include <QSerialPortInfo>
#include <QFileDialog>
#include <QDebug>
#include <QTime>
#include <QString>
#include <QTextStream>
#include <QMessageBox>
#include <QScrollBar>

#include "MainWindow.h"
#include "TryAction.h"

static const QBrush ansiPalette[] = {
    QBrush(QColor("#222")), QBrush(QColor("#C00")),
    QBrush(QColor("#2C0")), QBrush(QColor("#CC0")),
    QBrush(QColor("#00C")), QBrush(QColor("#C0C")),
    QBrush(QColor("#0CC")), QBrush(QColor("#CCC")),
    QBrush(QColor("#444")), QBrush(QColor("#F44")),
    QBrush(QColor("#4F4")), QBrush(QColor("#FF4")),
    QBrush(QColor("#44F")), QBrush(QColor("#F4F")),
    QBrush(QColor("#4FF")), QBrush(QColor("#FFF"))
};


ConsoleWindow::ConsoleWindow(QWidget *parent)
    : QMainWindow(parent)
    , ui(new Ui::ConsoleWindow)
{
    ui->setupUi(this);

    connect(ui->actionClear, &QAction::triggered, this, &ConsoleWindow::actionClear_triggered);
    connect(ui->actionSave, &QAction::triggered, this, &ConsoleWindow::actionSave_triggered);
    connect(ui->actionWrap, &QAction::triggered, this, &ConsoleWindow::actionWrap_triggered);

    connect(&ansiParser, &AnsiParser::attributesChanged, this, &ConsoleWindow::ansi_attributesChanged);
    connect(&ansiParser, &AnsiParser::printText, this, &ConsoleWindow::ansi_print_text);

    connect(&mainWindow->connection, &Connection::messageReceived, this, &ConsoleWindow::connection_messageReceived);
    connect(&timer, &QTimer::timeout, this, &ConsoleWindow::timer_timeout);

    cursor = QTextCursor(ui->plainTextEdit->document());

    auto p = ui->plainTextEdit->palette();
    p.setColor(QPalette::All, QPalette::Base, ansiPalette[0].color());
    p.setColor(QPalette::All, QPalette::Text, ansiPalette[7].color());
    ui->plainTextEdit->setPalette(p);

    auto font = ui->plainTextEdit->font();

#ifdef Q_OS_WIN
    font.setFamily("Lucida Console");
#else
    font.setFamily("Liberation Mono");
#endif

    ui->plainTextEdit->setFont(font);

    actionClear_triggered();
    timer.start(100);
}


ConsoleWindow::~ConsoleWindow()
{
    delete ui;
}


void ConsoleWindow::actionClear_triggered()
{
    ui->plainTextEdit->clear();
    ansi_attributesChanged(ansiParser.attributes);
}


void ConsoleWindow::actionSave_triggered()
{
    auto t = QDateTime::currentDateTime();

    auto fn = QString().sprintf(
        "log_%02d%02d%02d_%02d%02d%02d.txt",
        t.date().year() % 100, t.date().month(), t.date().day(),
        t.time().hour(), t.time().minute(), t.time().second()
    );

    auto modifiers = QApplication::keyboardModifiers();

    if (modifiers != Qt::ShiftModifier) {
        fn = QFileDialog::getSaveFileName(
            this, "Save console output", fn,
            "Text files (*.txt);;All Files (*.*)"
        );
    }

    if (fn == "")
        return;

    QFile f(fn);

    tryAction(
        [&]() { return f.open(QIODevice::WriteOnly); },
        [&]() { return QString("Can't save\n%1\n%2")
                        .arg(fn) .arg(f.errorString());
        }
    );

    f.write(ui->plainTextEdit->toPlainText().toLatin1());
    f.close();
}


void ConsoleWindow::actionWrap_triggered()
{
    if (ui->actionWrap->isChecked())
        ui->plainTextEdit->setLineWrapMode(QPlainTextEdit::WidgetWidth);
    else
        ui->plainTextEdit->setLineWrapMode(QPlainTextEdit::NoWrap);
}


void ConsoleWindow::connection_messageReceived(const msg_generic &msg)
{
    if (msg.h.id == MSG_ID_SHELL_TO_PC)
        rx_buf.append( QByteArray((char*)msg.data, msg.h.data_len) );
}


void ConsoleWindow::timer_timeout()
{
    if (ui->actionPause->isChecked())
        return;

    if (rx_buf.isEmpty())
        return;

    auto pte = ui->plainTextEdit;
    auto vsb = pte->verticalScrollBar();
    bool at_bottom = vsb->value() == vsb->maximum();

    pte->setUpdatesEnabled(false);

    ansiParser.parse(rx_buf);
    if (at_bottom)
        vsb->setValue(vsb->maximum());

    pte->setUpdatesEnabled(true);

    rx_buf.clear();
}


void ConsoleWindow::ansi_attributesChanged(const AnsiParser::Attributes &attr)
{
    unsigned fg = attr.foreground + (attr.bold ? 8 : 0);
    unsigned bg = attr.background;

    Q_ASSERT(fg < 16);
    Q_ASSERT(bg < 16);

    ansiFormat.setForeground(ansiPalette[fg]);
    ansiFormat.setBackground(ansiPalette[bg]);

    cursor.setCharFormat(ansiFormat);
}


void ConsoleWindow::ansi_print_text(const QString &text)
{
    cursor.insertText(text);
}
